### DISCUSSION TITLE
Does or doesn't GDScript have private variables/functions?

### DISCUSSION THREAD

**Discussion by miv391** [godotengine/godot-proposals#9985] [Scripting]
[Godot documentation says](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_styleguide.html#functions-and-variables) how you can make a private variable/function by prepending an underscore. However, to my understanding, the only functionality the underscore has, is that the variable/function is ignored by autocompletion. You can still access the variable/function from outside.

I propose that either
1) Variables/functions starting with underscore are changed to actually being private and not accessible from outside.
or
2) Documentation is changed to tell what the underscore actually does. It would be better not to use the word "private" at all to confuse users.

**Comment by filipworksdev** [comment]
Variables that start with _ don't autocomplete but are not private. Nothing is private in GDScript.

I actually just implemented `@private` annotation for testing purposes and yeah is pretty powerful in my testing gdscript branch https://github.com/goblinengine/gdscript2 in my testing/educational fork of Godot. Is not that hard to implement either.  
<img width="350" height="169" alt="image" src="https://github.com/user-attachments/assets/16ca11ee-032f-4ced-a36e-17e3c8e8d346" />
<img width="544" height="355" alt="image" src="https://github.com/user-attachments/assets/2b524587-ae50-4b01-af43-a009f364dcae" />
But as you can see _ is not really private just doesn't autocomplete.

**Reply by dalexeev** [reply]
> Is not that hard to implement either.

This issue has been raised several times in the Contributors Chat and on GitHub. The problem isn't that it's difficult to add the `private` keyword/annotation. The problem is that GDScript is a gradually typed language, so we can't just rely on static checks. Most static typing (with the exception of enumerations) provides runtime guarantees.

Proper implementation of this feature requires runtime overhead. However, this isn't the only difficulty. Some languages like PHP support access modifiers despite dynamic typing. However, the GDScript runtime, unlike the PHP runtime, doesn't control all code execution. GDScript relies heavily on the Godot core. For example, you can access custom variables and call custom methods through the `Object` API (`get()`, `set()`, `call()`), `Callable`, `Array` methods, etc.

When checking access modifiers, it's not just the **object** (whether a property or method is private) that's important, but also the **subject** attempting to access the property or method. Context is important, as most languages ​​allow access to private class members not only for the current instance but also for other instances of the same class.

Another issue is that in some cases, it may be desirable or necessary to allow access to private members from outside the class. For example, for the debugger, for the inspector, for assigning to exported variables, for calling private methods as signal callbacks, etc.

Currently, GDScript follows Python in this regard. The naming convention gracefully avoids runtime overhead and the complexity of context validation and accounting for necessary exceptions when accessing from outside. At the same time, the naming convention is quite clear and easy to use. If you access class members whose names begin with an underscore from the outside (`object._property = x`, `object._method()`), you should be sure of what you're doing. It just requires a bit of discipline (you should name private variables correctly).

An alternative solution that would allow us to limit ourselves to static checks is a warning system based on existing naming conventions. Warnings don't imply runtime guarantees, unlike keywords or annotations (that would be inconsistent with the rest of the language design).

* See https://github.com/godotengine/godot/pull/103768.

**Reply by miv391** [reply]
I think my original proposal number 2 would still be the best solution. Let's just stop calling the functions and variables starting with underscore private functions and variables. Instead, something like this would be better in the documentation: "Underscore is a convention to indicate to the programmer that the function/variable is not meant to be used outside the node/class. However, the GDScript language doesn't restrict that."

**Reply by dalexeev** [reply]
It's a matter of terminology. Yes, some users may mistakenly understand "private class members" for something other than what they actually are. However, we still need a short term for "private by convention underscore-prefixed class members." "Private" is the established term. Underscore-prefixed class members semantically function as private class members, even if they provide neither static nor runtime access restriction guarantees.[^1]

This is the current definition of "private class members" in GDScript, and it doesn't necessarily match definitions in other programming languages. Moreover, there's no single standard for access modifiers. Besides the fairly common `public`/`protected`/`private` arrangement, there are other options, with more or fewer access modifiers. For example, Haxe has only two modifiers: `public` and `private` (which is more like `protected`).

In GDScript, the term "private" is used in the warning name `UNUSED_PRIVATE_CLASS_VARIABLE`. The Godot editor also relies on this naming convention (for example, undocumented class members whose names begin with an underscore are not displayed in Editor Help).

[^1]: This is similar to the situation with JavaScript, which for a long time didn't have the `const` keyword. People simply used a naming convention (`var MY_CONSTANT = 42`) and agreed never to change uppercase-named variables (although technically there is no restriction on this). However, I'm sure that in many tutorials and discussions, such variables were referred to as "constants", not "conventionally constant uppercase-named variables."

**Reply by filipworksdev** [reply]
Hello everyone. In my implementation i actually block access to private  members that are outside the class both in the autocomplete/parser. I had to add a new property is_private in the node to do it but is just 1 bit flag I think.  You can see my POC implementation here is not that difficult to implement https://github.com/goblinengine/gdscript2/commit/671e5b9f4ea2167e6edceee11ba40ed6f380e8d1 but did i implement it correctly? I dont know. It also allowsd you to define the same variable/function name again if is set as private  in the parent class since it treats the private variables as if they no longer exist outside the class.

Edit: ignore the is_inline that was another tag I was working on and I eventually removed it because was not providing much performance benefit. The private implementation is even smaller than you see .

I think is a powerful feature and implemente as an annotaiton `@private` is opt in so doesn't really change the old syntax at all.

**Reply by dalexeev** [reply]
I haven't looked at your fork closely, but I mentioned in my comment above why proper implementation of access modifiers is more complex than it might first appear. GDScript is a gradually typed language. This means we strive to maintain parity between the language's dynamic and static sides, as well as full interoperability between them.

Access modifiers can't be implemented solely in the static analyzer or so that they only work in a static context. That is, if a user dynamically accesses a private member from another instance of the same class, it must work. Furthermore, we must account for many exceptional situations where access to private members must be possible even from outside (debugger, attaching a private method as a signal handler, etc.).

I'm not saying it's impossible to implement. I'm simply pointing out that all the pros and cons must be weighed, including runtime overhead and maintenance burden. Using naming conventions can be a reasonable compromise and a conscious design choice.
